### PR TITLE
MGMT-15073:Cluster reset - only system generated manifests to be deleted

### PR DESCRIPTION
**Pull request by paul-maidment**
Presently when a cluster is reset, we perform all sorts of awkward workarounds to exclude files from deletion, when in reality all we should do is delete only the system generated manifests, leaving all other files and logs intact.

Moreover, when performing a permenant deletion, ALL files should be deleted.

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/cluster/cluster.go [hunk 1]
```diff
@@ -1399,32 +1403,106 @@ func (m *Manager) SetConnectivityMajorityGroupsForCluster(clusterID strfmt.UUID,
...
-		return common.NewApiError(
+		return nil, common.NewApiError(
 			http.StatusInternalServerError,
 			errors.Errorf(msg))
 	}
+	return files, err
+}
 
-	var failedToDelete []string
+func (m *Manager) ResetClusterFiles(ctx context.Context, c *common.Cluster, objectHandler s3wrapper.API) error {
```
Comment (carbonin) [inline internal/cluster/cluster.go:1430]: This function is only dealing with manifests, right? Maybe we should name it that way.

File: internal/cluster/cluster.go [hunk 2]
```diff
@@ -1399,32 +1403,106 @@ func (m *Manager) SetConnectivityMajorityGroupsForCluster(clusterID strfmt.UUID,
...
+			isSystemGeneratedManifest := true
+			for _, manifest := range userManifests {
+				if manifest.Folder == currentFolder && manifest.FileName == currentFile {
+					isSystemGeneratedManifest = false
+					break
+				}
+			}
+			if isSystemGeneratedManifest {
+				log.Debugf("Deleting cluster %s S3 file: %s", c.ID.String(), file)
+				_, err := objectHandler.DeleteObject(ctx, file)
```
Comment (carbonin) [inline internal/cluster/cluster.go:1454]: Where do we handle deleting the metadata? Or is that only created for the user manifests? I forget which direction we ended up going on that one.

File: internal/cluster/cluster.go [hunk 3]
```diff
@@ -1399,32 +1403,106 @@ func (m *Manager) SetConnectivityMajorityGroupsForCluster(clusterID strfmt.UUID,
...
 
-	var failedToDelete []string
+func (m *Manager) ResetClusterFiles(ctx context.Context, c *common.Cluster, objectHandler s3wrapper.API) error {
+	log := logutil.FromContext(ctx, m.log)
+	files, err := m.getClusterFilesForFolder(ctx, c, objectHandler, "")
+	if err != nil {
+		return err
+	}
+	// ListClusterManifestsInternal only returns user generated manifests so we can use this to determine whether or not a manifest should be deleted on reset.
+	userManifests, err := m.manifestApi.ListClusterManifestsInternal(ctx, manifests.V2ListClusterManifestsParams{ClusterID: *c.Cluster.ID})
```
Comment (carbonin) [inline internal/cluster/cluster.go:1437]: It seems like it would be easier to create a manifestsApi function to return the system manifests then just delete them directly.

This strategy of getting the user manifests then iterating through all the files eliminating those is strange. Is there any other reason you're doing it this way?

File: internal/cluster/cluster.go [hunk 4]
```diff
@@ -1399,32 +1403,106 @@ func (m *Manager) SetConnectivityMajorityGroupsForCluster(clusterID strfmt.UUID,
 	return m.setConnectivityMajorityGroupsForClusterInternal(cluster, db)
 }
 
-func (m *Manager) deleteClusterFiles(ctx context.Context, c *common.Cluster, objectHandler s3wrapper.API, folder string) error {
-	log := logutil.FromContext(ctx, m.log)
+func (m *Manager) extractManifestFolderAndFilename(ctx context.Context, path string) (string, string, bool) {
```
Comment (carbonin) [inline internal/cluster/cluster.go:1406]: I think making this a function is actually a bit more confusing than just embedding this in the only place it's used.

File: internal/cluster/cluster.go [hunk 5]
```diff
@@ -1399,32 +1403,106 @@ func (m *Manager) SetConnectivityMajorityGroupsForCluster(clusterID strfmt.UUID,
...
+	if err != nil {
+		return err
+	}
 	for _, file := range files {
-		//skip log and manifests deletion when deleting cluster files
-		if folder == "" && (strings.Contains(file, "logs") || strings.Contains(file, constants.ManifestFolder) || strings.Contains(file, constants.ManifestMetadataFolder)) {
-			continue
+		// Path is in the form <cluster-id>/manifests/<folder-name>/<file-name>
+		currentFolder, currentFile, isManifest := m.extractManifestFolderAndFilename(ctx, file)
+		if isManifest {
```
Comment (carbonin) [inline internal/cluster/cluster.go:1444]: Was there previously anything else in the `""` folder other than manifests? It seems like we would not be deleting that now, but would have been previously.

File: internal/cluster/cluster.go [hunk 6]
```diff
@@ -1399,32 +1403,106 @@ func (m *Manager) SetConnectivityMajorityGroupsForCluster(clusterID strfmt.UUID,
...
+	}
+	return nil
+}
+
+func (m *Manager) DeleteClusterLogs(ctx context.Context, c *common.Cluster, objectHandler s3wrapper.API) error {
+	return m.deleteClusterFilesInFolders(ctx, c, objectHandler, []string{"logs"})
+}
+
+func (m *Manager) deleteAllClusterFiles(ctx context.Context, c *common.Cluster, objectHandler s3wrapper.API) error {
+	folders := []string{"", "logs", constants.ManifestFolder, constants.ManifestMetadataFolder}
```
Comment (carbonin) [inline internal/cluster/cluster.go:1469]: Does `""` not contain all the other entries? If we know we're going to delete everything do we really need all the other specific folders?

File: internal/cluster/cluster_test.go [hunk 7]
```diff
@@ -3172,6 +3179,9 @@ var _ = Describe("Permanently delete clusters", func() {
 		Expect(db.First(&common.Cluster{}, "id = ?", c2.ID).RowsAffected).Should(Equal(int64(0)))
 		Expect(db.Unscoped().First(&common.Cluster{}, "id = ?", c2.ID).RowsAffected).Should(Equal(int64(1)))
 
+		// ListClusterManifestsInternal(ctx, manifests.V2ListClusterManifestsParams{ClusterID: *c.Cluster.ID})
```
Comment (carbonin) [inline internal/cluster/cluster_test.go:3182]: You can remove this comment.

**Review by carbonin — CHANGES_REQUESTED** [review]
This implementation seems more complicated than the previous one to me.
We have a few instances of nested loops that I think we can remove and the logic to determine if a manifest should be deleted on reset seems inverted to me.

File: internal/cluster/cluster.go (hunk 2 above)
Comment (paul-maidment) [reply internal/cluster/cluster.go:1454]: Metadata is only created for user manifests

File: internal/cluster/cluster.go (hunk 3 above)
Comment (paul-maidment) [reply internal/cluster/cluster.go:1437]: What about an API function to delete system manifests directly via API, maybe that would work better?

File: internal/cluster/cluster.go (hunk 6 above)
Comment (paul-maidment) [reply internal/cluster/cluster.go:1469]: Old behaviour did this too, did not quite understand why, I am just replicating here but was curious about this too.

File: internal/cluster/cluster_test.go [hunk 8]
```diff
@@ -3172,6 +3179,9 @@ var _ = Describe("Permanently delete clusters", func() {
...
+		manifestsAPI.EXPECT().ListClusterManifestsInternal(gomock.Any(), manifests.V2ListClusterManifestsParams{ClusterID: *c1.Cluster.ID}).AnyTimes()
+		manifestsAPI.EXPECT().ListClusterManifestsInternal(gomock.Any(), manifests.V2ListClusterManifestsParams{ClusterID: *c2.Cluster.ID}).AnyTimes()
 		mockS3Api.EXPECT().DeleteObject(gomock.Any(), c1.ID.String()).Return(false, nil).Times(1)
```
[the remark this answers is not in the document]
Comment (paul-maidment) [reply internal/cluster/cluster_test.go:3185]: Because these functions are listing user manifests since the change.
Each file from the directory is then checked against the list of known user manifests as follows

If the file is a manifest and is not a user manifest then it is a system manifest.
In the case of legacy clusters (ones from before the "manifest metadata" change) the files are always considered to be "user manifests"

File: internal/manifests/manifests_test.go [hunk 9]
```diff
@@ -791,6 +792,30 @@ spec:
 			Expect(responsePayload.Payload.Folder).To(Equal(defaultFolder))
 		})
 	})
+
+	Context("Manifest path parsing functions", func() {
```
Comment (carbonin) [inline internal/manifests/manifests_test.go:796]: Use `Describe`  with the function you're testing and `It` for individual behaviors you're testing. So something like

```go
Describe("ParsePath", func() {
  It("parses a path into folder and filename", func() {
    ...
  })
  It("fails when provided a non-manifest path", func() {
    ..
  })
})
Describe("IsManifest", func() {
  ...
})
```

This way a test failure will show exactly which function is failing and what about it isn't working.

File: internal/manifests/manifests.go [hunk 10]
```diff
@@ -327,7 +338,8 @@ func GetManifestObjectName(clusterID strfmt.UUID, fileName string) string {
 
 // GetManifestObjectName returns the manifest object name as stored in S3
 func GetManifestMetadataObjectName(clusterID strfmt.UUID, fileName string, manifestSource string) string {
-	return filepath.Join(string(clusterID), constants.ManifestMetadataFolder, fileName, manifestSource)
+	objectName := filepath.Join(string(clusterID), constants.ManifestMetadataFolder, fileName, manifestSource)
+	return objectName
```
Comment (carbonin) [inline internal/manifests/manifests.go:342]: Why was this changed?

Comment (paul-maidment) [reply internal/manifests/manifests.go:342]: Mainly debugging, will revert it when I commit.

File: internal/cluster/cluster.go [hunk 11]
```diff
@@ -1399,32 +1403,47 @@ func (m *Manager) SetConnectivityMajorityGroupsForCluster(clusterID strfmt.UUID,
...
 		m.log.WithError(err).Errorf(msg)
-		return common.NewApiError(
+		return nil, common.NewApiError(
 			http.StatusInternalServerError,
 			errors.Errorf(msg))
 	}
+	return files, err
+}
 
+func (m *Manager) ResetClusterFiles(ctx context.Context, c *common.Cluster, objectHandler s3wrapper.API) error {
```
Comment (carbonin) [inline internal/cluster/cluster.go:1419]: I think we should add unit tests for this.

File: internal/cluster/cluster_test.go [hunk 12]
```diff
@@ -3985,3 +3993,73 @@ var _ = Describe("UploadEvents", func() {
...
+		mockObjectHandler = s3wrapper.NewMockAPI(ctrl)
+		capi = NewManager(getDefaultConfig(), common.GetTestLog(), db, commontesting.GetDummyNotificationStream(ctrl), nil, nil, nil, nil, nil, dummy, mockOperators, nil, mockObjectHandler, nil, nil, mockManifestsApi)
+	})
+
+	AfterEach(func() {
+		common.DeleteTestDB(db, dbName)
+		ctrl.Finish()
+	})
+
+	It("Should clear only user generated manifests when performing a reset", func() {
```
Comment (carbonin) [inline internal/cluster/cluster_test.go:4042]: Isn't this test name incorrect?
It should clear only system manifests when resetting, right?

Comment (paul-maidment) [reply internal/cluster/cluster_test.go:4042]: Yes, will change this

**Review by carbonin — APPROVED** [review]
